### PR TITLE
JDBetteridge/python312

### PR DESCRIPTION
**Pull request by JDBetteridge**
# Description
This is now ready to land, once numpy 2.0 is released on 16th June

https://numpy.org/news/#numpy-20-release-date-june-16

Linked PRs:
- ~[FIAT](https://github.com/firedrakeproject/fiat/pull/72)~ DONE
- ~[PyOP2](https://github.com/OP2/PyOP2/pull/721)~ DONE
- ~[TSFC](https://github.com/firedrakeproject/tsfc/pull/313)~ DONE
- ~[Thetis](https://github.com/thetisproject/thetis/pull/364)~ DONE
- ~[TinyASM](https://github.com/florianwechsung/TinyASM/pull/13)~ We want to incorperate this into Firedrake, rather than merge this PR, see [3604](https://github.com/firedrakeproject/firedrake/pull/3604) instead DONE

**Code Coverage**

Code coverage can be re-enabled when Ubuntu 24.04 updates to Python 3.12.4, see https://github.com/python/cpython/issues/119118 for the performance regression

### CODE REVIEW

File: scripts/firedrake-install
Comment (JDBetteridge) [inline scripts/firedrake-install:1299]:
```suggestion
            check_call([pip, "install", "-e", "."])
```

File: .github/workflows/build.yml
Comment (JDBetteridge) [inline .github/workflows/build.yml:29]:
```suggestion
      image: firedrakeproject/firedrake-env:latest
```

Wait until ready to merge

File: .github/workflows/build.yml
Comment (JDBetteridge) [inline .github/workflows/build.yml:88]:
```suggestion
```
Wait until ready to merge

File: scripts/firedrake-install
Comment (JDBetteridge) [inline scripts/firedrake-install:1725]:
```suggestion
run_pip(["install", "-U", "numpy"])
```
Wait until ready to merge

Comment (JDBetteridge) [inline scripts/firedrake-install:1725]:
```suggestion
run_pip(["install", "-U", "numpy"])
```

File: firedrake/cython/dmcommon.pyx
Comment (ksagiyam) [inline firedrake/cython/dmcommon.pyx:2556]: I think you did not intend to remove `const`s here?

File: .github/workflows/build.yml
Comment (JDBetteridge) [inline .github/workflows/build.yml:164]:
```suggestion
```

File: firedrake/cython/dmcommon.pyx
Comment (JDBetteridge) [reply firedrake/cython/dmcommon.pyx:2556]: @connorjward is that right? I think you made this commit

Comment (connorjward) [reply firedrake/cython/dmcommon.pyx:2556]: I don't really remember exactly why I did this. I tackled a bunch of signature changes because Cython 3 is a little stricter. I think it's likely that this was a necessary change but I'm not 100% sure.

Comment (ksagiyam) [reply firedrake/cython/dmcommon.pyx:2556]: This is something explicitly fixed in https://github.com/firedrakeproject/firedrake/pull/3503/files, so, maybe, when Connor started making Cython3 changes, this fix was not in, and later merge/rebase process somehow dropped `const`s.

Comment (connorjward) [reply firedrake/cython/dmcommon.pyx:2556]: Ah I see. We both modified the same line. Good catch.

**Comment by rckirby** [comment]
FYI, I went ahead and did the FIAT merge.   Nothing was changed pointing to various numpy versions.  All it had was `np.factorial` replaced by `math.factorial` which won't break anything.

File: scripts/firedrake-install
Comment (dham) [inline scripts/firedrake-install:37]:
```suggestion
               "git+ssh://github.com/thetisproject/thetis.git#egg=thetis"),
```

File: scripts/firedrake-install
Comment (dham) [inline scripts/firedrake-install:1340]: try reverting this.

**Review by dham — DISMISSED** [review]
rebase and lose branches of pytorch and thetis.

**Review by connorjward — CHANGES_REQUESTED** [review]
please update page about 3.8 and also mumble mumble ARM support (to badly quote @JDBetteridge)

**Review by connorjward — DISMISSED** [review]
This all looks good to me!

**Pull request merged**
